From 0ad5f547eba5696439592c6fd20e5146930e6c63 Mon Sep 17 00:00:00 2001
From: Steve Loughran <stevel@apache.org>
Date: Thu, 20 Oct 2016 14:50:30 +0100
Subject: [PATCH 2163/2848] HADOOP-13735
 ITestS3AFileContextStatistics.testStatistics()
 failing. Contributed by Pieter Reuse

(cherry picked from commit 9ae270af02c243993f853513c731cb268430e492)

Change-Id: Ib0cba5bebdb2f99a644927e8188c750f9de26022
---
 .../org/apache/hadoop/fs/s3a/S3AFileSystem.java    |    2 +-
 .../apache/hadoop/fs/s3a/S3AInstrumentation.java   |   13 ++++++++++---
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
index fe34d3d..2177a5a 100644
--- a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
+++ b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java
@@ -561,7 +561,7 @@ public FSDataOutputStream create(Path f, FsPermission permission,
               progress,
               partSize,
               blockFactory,
-              instrumentation.newOutputStreamStatistics(),
+              instrumentation.newOutputStreamStatistics(statistics),
               new WriteOperationHelper(key)
           ),
           null);
diff --git a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInstrumentation.java b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInstrumentation.java
index 963c53f..fb8c852 100644
--- a/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInstrumentation.java
+++ b/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInstrumentation.java
@@ -37,6 +37,7 @@
 import java.util.Map;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicLong;
+import org.apache.hadoop.fs.FileSystem.Statistics;
 
 import static org.apache.hadoop.fs.s3a.Statistic.*;
 
@@ -639,9 +640,8 @@ public String toString() {
    * Create a stream output statistics instance.
    * @return the new instance
    */
-
-  OutputStreamStatistics newOutputStreamStatistics() {
-    return new OutputStreamStatistics();
+  OutputStreamStatistics newOutputStreamStatistics(Statistics statistics) {
+    return new OutputStreamStatistics(statistics);
   }
 
   /**
@@ -677,6 +677,12 @@ private void mergeOutputStreamStatistics(OutputStreamStatistics statistics) {
     private final AtomicLong queueDuration = new AtomicLong(0);
     private final AtomicLong exceptionsInMultipartFinalize = new AtomicLong(0);
 
+    private Statistics statistics;
+
+    public OutputStreamStatistics(Statistics statistics){
+      this.statistics = statistics;
+    }
+
     /**
      * Block is queued for upload.
      */
@@ -717,6 +723,7 @@ void blockUploadFailed(long duration, int blockSize) {
     /** Intermediate report of bytes uploaded. */
     void bytesTransferred(long byteCount) {
       bytesUploaded.addAndGet(byteCount);
+      statistics.incrementBytesWritten(byteCount);
       bytesPendingUpload.addAndGet(-byteCount);
       incrementGauge(STREAM_WRITE_BLOCK_UPLOADS_DATA_PENDING, -byteCount);
     }
-- 
1.7.9.5

